crates/bevy_ecs/src/query/iter.rs

```diff
@@ -1841,7 +2650,7 @@ impl<'w, 's, D: QueryData, F: QueryFilter> QueryIterationCursor<'w, 's, D, F> {
     }
 
     // NOTE: If you are changing query iteration code, remember to update the following places, where relevant:
-    // QueryIter, QueryIterationCursor, QuerySortedIter, QueryManyIter, QueryCombinationIter, QueryState::par_fold_init_unchecked_manual
+    // QueryIter, QueryIterationCursor, QuerySortedIter, QueryManyIter, QuerySortedManyIter, QueryCombinationIter, QueryState::par_fold_init_unchecked_manual
```

Comment:
You got it! More documentation should definitely be made. Currently, the docs for `Query` do not mention sorting, which they really should as well.

Yeah, there are a lot of iterator types, however it is normal in the rust ecosystem to have different structs for different forms of iteration, just like every adapter on `iterator` returns its own struct. (That bevy has combined mutable and immutable into a single type is already unusual)

However I do not think combining them is a good idea. That would not reduce code duplication by much, since different forms of iteration will still need their own impl blocks regardless.
Quite the opposite, I would find it more confusing both from a dev, and the user standpoint.
Since picking the "wrong kind" of iteration is quite common, and iterators like to nest, it is very important to be as clear as we can inside errors. (Errors unpack type aliases)
Besides, future iterator types that actually do just do the same kind of iteration can be type aliases in clear manner:
`pub type QuerySortedManyUniqueIter<...> = QuerySortedIter<...>;`

Reply:
My concern with duplication is mainly the various `fetch_next` methods, which duplicate the logic of `Query::get`.  `QueryManyIter::fetch_next_aliased_unchecked`, `QuerySortedIter::fetch_next`, and `QuerySortedManyIter::fetch_next_aliased_unchecked` are almost identical, and it's hard to see what the differences are if you want to make changes to the logic.  

If we want to have multiple public types, then maybe we can have a single inner type with the complicated `fetch_next` implementation, and have the other types wrap that with appropriate safety assertions.  

But we can always refactor once this and #13477 are both merged, so I don't think we need to worry about it yet!

Reply:
There are several properties about their entities query iterators can rely on/guarantee:

Iterator | Uniqueness | Existence | QF eval* | dense EL/A order**
--- | --- | --- | --- | --- 
QueryManyIter | Unknown | Unknown | Unknown | Unknown
QueryManyUniqueIter | Known | Unknown | Unknown | Unknown
QuerySortedManyIter | Unknown | Known | Known | Depends
QuerySortedManyUniqueIter | Known | Known | Known | Depends
QuerySortedIter | Known | Known | Known | Depends
QueryIter | Known | Known | Depends | Known 
QueryParIter | Known | Known | Depends | n/a?
QueryParManyIter | Unknown | Unknown | Unknown | n/a?
QueryParManyUniqueIter | Known | Unknown | Unknown | n/a?
QueryCombinationsIter | n/a | Known | Depends | n/a
QuerySortedCombinationsIter | n/a | Known | Known | n/a
QueryCombinationsManyIter | n/a | Unknown | Unknown | n/a
QuerySortedCombinationsManyIter | n/a | Known | Known | n/a

*non-archetypal QueryFilters
**dense EntityLocation/Archetype order

Take this as a rough sketch, it is my current understanding of sensible iterator types, given current/potential API. 
It is not an exhaustive or definite list, I am also least familiar with combination/parallel iterators)
Each of these properties can change what `fetch_next`/its equivalent can do/skip, which makes it really difficult to try and deduplicate without introducing unnecessary complexity or perf hits. I'm not gonna say it is impossible, but in the meantime, documenting these properties would likely make more sense imo.